src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -524,11 +528,35 @@ public void OnHeader(Span<byte> name, Span<byte> value)
             HttpRequestHeaders.Append(name, value);
         }
 
+        public void OnTrailer(Span<byte> name, Span<byte> value)
+        {
+            // Trailers still count towards the limit.
+            _requestHeadersParsed++;
+            if (_requestHeadersParsed > ServerOptions.Limits.MaxRequestHeaderCount)
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.TooManyHeaders);
+            }
+
+            string key = name.GetHeaderName();
+            var valueStr = value.GetAsciiOrUTF8StringNonNullCharacters();
+            RequestTrailers.Append(key, valueStr);
```

Comment:
Why not pass the spans directly to Append like OnHeader?

Reply:
RequestTrailers is a vanilla IHeaderDictionary, not a custom collection. It's not clear that most of the header optimizations we've done elsewhere are relevant to request trailers.